modules/tutorial-advanced-minimalist-fulcro/pages/index.adoc

```diff
@@ -0,0 +1,338 @@
+
+==== On router timeouts and errors
+
+What happens when loading data in a deferred route takes too long? What if it fails? The router has you covered:
+
+.Router with timeouts, a loading indicator, and an error message
+====
+```clojure
+(dr/defrouter MyRouter [_ {:keys [current-state route-factory route-props]}]
+  {:router-targets [AllPeople Person]}
+  ;; The body of the router is displayed only when the target is not ready,
+  ;; i.e. in one of the states below (unless you set `:always-render-body?`)
+  (case current-state ; <1>
+    nil (println "MISTAKE: MyRouter is displayed but has never been routed to yet") ; <2>
+    :pending (dom/div "Loading...") ; <3>
+    :failed (dom/div "Failed!")     ; <4>
+    (println "Should never come here:" current-state))) ; <5>
+```
+<1> The body of the router is only displayed when the target is not ready (unless you set `:always-render-body? true`). Thus the only router states we will ever see here are those listed below.
+<2> `current-state = nil` is the initial state, before the router has been routed to - i.e. when you failed to initialize it properly via `change-route!`, as described above (there is also the `:inital` state but I have not seen it in practice)
+<3> `:pending` is active for a deferred route, after the `:deferred-timeout` passed but before the `:error-timeout` - it also ends if `dr/target-ready` is transacted for the target ident
+<4> `:failed` if the deferred route has not become ready (as indicated by `dr/target-ready`) within the `:error-timeout` since the route event
+<5> We do not expect to see any other states, namely `:routed` - because in this state the target itself is displayed and not the router's body (unless we set `:always-render-body? true`)
+====
+
+Key points to know:
+
+* There are two timeouts that control the router's behavior with respect to deferred routes, `:deferred-timeout` (default 20ms)
+  and `:error-timeout` (default 5s). You can override the default values when calling `dr/change-route!`, link:{url-cljdoc-fulcro}com.fulcrologic.fulcro.routing.dynamic-routing#change-route![see its docstring].
+* You can use the body of the router and its `current-state`. It is only displayed when the target is not "ready" to be displayed
+* The router switches from displaying the previous target to displaying the router's body while waiting for a pending target only after the `:deferred-timeout` has passed (but it switches to the target immediately when it gets ready)
+* If the mutation `(dr/target-ready {:target <ident>}` is not triggered within `:error-timeout` of `change-route!` then the state becomes `:failed` (the target can still become "ready" later, if the load eventually finishes, and the target will be displayed)
+* Thus we can use `:pending` to display a loading indicator, if it takes longer than the `:deferred-timeout`. We can use the `:failed` state to show an error message telling the user it is either taking too long or failed
+
+=== Nesting routers
```

Comment:
Hukka: what happens with nested routers and partial load failures (especially with multiple remotes). That is, how to do the stuff that React Suspense does

JH: No content below the component that failed to load its data will be displayed. How could it, if the component lacks its data and thus cannot display its body? Does that answer your question?

Reply:
I mean that in vanilla React the routing, and the loading with Suspense levels are distinct concerns, and the developer can choose the boundaries where loading spinners and errors are shown, separately from which parts of the UI tree fork based on the URL, for example.

Is it somehow possible to control nested router loading/error behavior with Fulcro? Not having tried it yet, I'm not even sure if it is the innermost/closest router that would show the loading/error, or the topmost (since the tree under it didn't load completely).

Reply:
It is up to you whether you want to use :will-enter for loading data or not. You can decide not to use that on the nested routers and instead load all the data for all the nested pages inside the top router's :will-enter. You have all the info you need there, namely the full route. You *can* interwine routing/displaying and loading but you *don't have to* do it. You are free to load! where you want and handle errors how you want.

Does that help?